File: python/cadreen/resources/traces.py
from __future__ import annotations

from typing import Any, Optional

from ..client import HttpClient
from ..resources.intent import _parse_intelligence
from ..types import (
    IntelligenceTraceEntry,
    ListIntelligenceResponse,
    IntelligenceStats,
    Pagination,
    ReplayResult,
    HandoffPacket,
    PromoteResult,
)


class TracesResource:
    def __init__(self, client: HttpClient) -> None:
        self._client = client

    async def get(self, id: str) -> IntelligenceTraceEntry:
        raw = await self._client.get(f"/api/v1/cadreen/intelligence/{id}")
        entry = IntelligenceTraceEntry(
            id=raw["id"],
            domain=raw.get("domain", ""),
            request_path=raw.get("request_path", ""),
            request_method=raw.get("request_method", ""),
            meta=_parse_intelligence(raw.get("meta", {})),
            created_at=raw.get("created_at"),
        )
        return entry

    async def list(
        self,
        domain: Optional[str] = None,
        decision: Optional[str] = None,
        from_: Optional[str] = None,
        to: Optional[str] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
    ) -> ListIntelligenceResponse:
        params: dict[str, Any] = {}
        if domain is not None:
            params["domain"] = domain
        if decision is not None:
            params["decision"] = decision
        if from_ is not None:
            params["from"] = from_
        if to is not None:
            params["to"] = to
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset

        raw = await self._client.get("/api/v1/cadreen/intelligence", params)
        traces = [
            IntelligenceTraceEntry(
                id=t["id"],
                domain=t.get("domain", ""),
                request_path=t.get("request_path", ""),
                request_method=t.get("request_method", ""),
                meta=_parse_intelligence(t.get("meta", {})),
                created_at=t.get("created_at"),
            )
            for t in raw.get("traces", [])
        ]
        pagination = None
        if raw.get("pagination"):
            pagination = Pagination(
                limit=raw["pagination"]["limit"],
                offset=raw["pagination"]["offset"],
                has_more=raw["pagination"]["has_more"],
            )
        return ListIntelligenceResponse(
            traces=traces,
            count=raw.get("count", 0),
            pagination=pagination,
        )

    async def stats(self) -> IntelligenceStats:
        raw = await self._client.get("/api/v1/cadreen/intelligence/stats")
        return IntelligenceStats(
            traces_24h=raw.get("traces_24h", 0),
            traces_7d=raw.get("traces_7d", 0),
            traces_30d=raw.get("traces_30d", 0),
            avg_confidence_by_domain=raw.get("avg_confidence_by_domain", {}),
            gap_detection_rate=raw.get("gap_detection_rate", 0.0),
            governance_decisions=raw.get("governance_decisions", {}),
        )

    async def replay(self, id: str, mode: str = "current") -> ReplayResult:
        raw = await self._client.post(f"/api/v1/cadreen/intelligence/{id}/replay", {"mode": mode})
        return ReplayResult(
            trace_id=raw["trace_id"],
            mode=raw.get("mode", ""),
            domain=raw.get("domain", ""),
            original_gate=raw.get("original_gate", ""),
            original_confidence=raw.get("original_confidence", 0.0),
            current_gate=raw.get("current_gate", ""),
            current_confidence=raw.get("current_confidence", 0.0),
            gate_changed=raw.get("gate_changed", False),
            change_summary=raw.get("change_summary", ""),
            current_capability=raw.get("current_capability", {}),
            current_memory=raw.get("current_memory", {}),
            current_gaps=raw.get("current_gaps", {}),
            replay_note=raw.get("replay_note", ""),
        )

    async def handoff(self, id: str) -> HandoffPacket:
        raw = await self._client.get(f"/api/v1/cadreen/intelligence/{id}/handoff")
        return HandoffPacket(
            trace_id=raw["trace_id"],
            domain=raw.get("domain", ""),
            created_at=raw.get("created_at", ""),
            governance=raw.get("governance", {}),
            what_the_system_knew=raw.get("what_the_system_knew", {}),
            what_the_system_didnt_know=raw.get("what_the_system_didnt_know", {}),
            what_happened=raw.get("what_happened", {}),
            suggested_actions=raw.get("suggested_actions", []),
            next_action=raw.get("next_action", {}),
            trace_url=raw.get("trace_url", ""),
        )

    async def promote(self, id: str, kind: str, name: str | None = None) -> PromoteResult:
        body: dict[str, Any] = {"kind": kind}
        if name:
            body["name"] = name
        raw = await self._client.post(f"/api/v1/cadreen/intelligence/{id}/promote", body)
        return PromoteResult(
            id=raw["id"],
            kind=raw.get("kind", ""),
            status=raw.get("status", ""),
            tool_name=raw.get("tool_name"),
            tool_sequence=raw.get("tool_sequence"),
            source_trace_id=raw.get("source_trace_id", ""),
        )

File: typescript/src/resources/traces.ts
import type {
  IntelligenceTraceEntry,
  ListIntelligenceResponse,
  IntelligenceStats,
  TraceExplain,
  ReplayResult,
  HandoffPacket,
  PromoteResult,
} from "../types";
import { HttpClient } from "../client";

export class TracesResource {
  constructor(private client: HttpClient) {}

  async get(id: string): Promise<IntelligenceTraceEntry & { explain: () => TraceExplain }> {
    const trace = await this.client.get<IntelligenceTraceEntry>(
      `/api/v1/cadreen/intelligence/${encodeURIComponent(id)}`
    );
    return {
      ...trace,
      explain: (): TraceExplain => {
        const meta = trace.meta;
        const steps: string[] = [];
        if (meta.capability.total_available > 0) {
          steps.push(`${meta.capability.healthy_count}/${meta.capability.total_available} capabilities healthy`);
        }
        if (meta.governance.active) {
          steps.push(`Governance: ${meta.governance.decision || "active"} (confidence: ${meta.governance.confidence || 0})`);
        }
        if (meta.humility.gaps_detected && meta.humility.gaps_detected > 0) {
          steps.push(`${meta.humility.gaps_detected} gaps detected (${meta.humility.blocking || 0} blocking)`);
        }
        if (meta.memory.knowledge_queried) {
          steps.push(`${meta.memory.knowledge_queried} knowledge items queried`);
        }
        const recommendations: string[] = [];
        if (meta.humility.blocking && meta.humility.blocking > 0) {
          recommendations.push("Resolve blocking capability gaps before proceeding");
        }
        if (!meta.memory.healthy) {
          recommendations.push("Knowledge store is degraded — expect reduced context quality");
        }
        return {
          summary: meta.summary || `Trace ${trace.id}: ${steps.join("; ")}`,
          steps,
          recommendations: recommendations.length > 0 ? recommendations : undefined,
        };
      },
    };
  }

  async list(options?: {
    domain?: string;
    decision?: string;
    from?: string;
    to?: string;
    limit?: number;
    offset?: number;
  }): Promise<ListIntelligenceResponse> {
    const params: Record<string, string | number | undefined> = {};
    if (options?.domain) params.domain = options.domain;
    if (options?.decision) params.decision = options.decision;
    if (options?.from) params.from = options.from;
    if (options?.to) params.to = options.to;
    if (options?.limit) params.limit = options.limit;
    if (options?.offset) params.offset = options.offset;
    return this.client.get<ListIntelligenceResponse>("/api/v1/cadreen/intelligence", params);
  }

  async stats(): Promise<IntelligenceStats> {
    return this.client.get<IntelligenceStats>("/api/v1/cadreen/intelligence/stats");
  }

  async replay(id: string, mode?: "current" | "historical"): Promise<ReplayResult> {
    return this.client.post<ReplayResult>(`/api/v1/cadreen/intelligence/${encodeURIComponent(id)}/replay`, {
      mode: mode || "current",
    });
  }

  async handoff(id: string): Promise<HandoffPacket> {
    return this.client.get<HandoffPacket>(`/api/v1/cadreen/intelligence/${encodeURIComponent(id)}/handoff`);
  }

  async promote(id: string, kind: "healing" | "memory" | "procedure", name?: string): Promise<PromoteResult> {
    return this.client.post<PromoteResult>(`/api/v1/cadreen/intelligence/${encodeURIComponent(id)}/promote`, {
      kind,
      name,
    });
  }
}
